src/LibraryManager.Vsix/Shared/VsHelpers.cs

```diff
@@ -336,5 +372,138 @@ public static Project GetDTEProjectFromConfig(string file)
 
             return null;
         }
+
+        private static async Task<bool> AddFilesToHierarchyAsync(IVsHierarchy hierarchy, IEnumerable<string> filePaths, Action<string, LogLevel> logAction, CancellationToken cancellationToken)
+        {
+            int batchSize = 10;
+
+            List<string> filesToAdd = filePaths.ToList();
+
+            while (filesToAdd.Any())
+            {
+                List<string> nextBatch = filesToAdd.Take(batchSize).ToList();
+                bool success = await AddProjectItemsInBatchAsync(hierarchy, nextBatch, logAction, cancellationToken);
+
+                if (!success)
+                {
+                    return false;
+                }
+
+                await System.Threading.Tasks.Task.Yield();
+
+                int countToDelete = filesToAdd.Count() >= batchSize ? batchSize : filesToAdd.Count();
+                filesToAdd.RemoveRange(0, countToDelete);
+            }
+
+            return true;
+        }
+
+        private static async Task<bool> AddProjectItemsInBatchAsync(IVsHierarchy vsHierarchy, List<string> filePaths, Action<string, LogLevel> logAction, CancellationToken cancellationToken)
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            IVsProjectBuildSystem bldSystem = vsHierarchy as IVsProjectBuildSystem;
+
+            try
+            {
+                if (bldSystem != null)
+                {
+                    bldSystem.StartBatchEdit();
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var vsProject = (IVsProject)vsHierarchy;
+                VSADDRESULT[] result = new VSADDRESULT[filePaths.Count()];
+
+                vsProject.AddItem(VSConstants.VSITEMID_ROOT,
+                            VSADDITEMOPERATION.VSADDITEMOP_LINKTOFILE,
+                            string.Empty,
+                            (uint)filePaths.Count(),
+                            filePaths.ToArray(),
+                            IntPtr.Zero,
+                            result);
+
+                foreach (string filePath in filePaths)
+                {
+                    logAction.Invoke(string.Format(Resources.Text.LibraryAddedToProject, filePath.Replace('\\', '/')), LogLevel.Operation);
+                }
+            }
+            catch (Exception)
+            {
+                logAction.Invoke(string.Format(Resources.Text.FailToAddLibrariesToProject), LogLevel.Operation);
```

Comment:
Is it possible to report which succeeded and failed?  Or does the vsProject commit the batch atomically?